lisa/_kmod.py

```diff
@@ -2267,4 +2286,40 @@ def from_target(cls, target, **kwargs):
     def _event_features(cls, events):
         return set(f'event__{event}' for event in events)
 
+    def install(self, kmod_params=None):
+        try:
+            self.uninstall()
+        except Exception:
+            pass
+
+        busybox = quote(self.target.busybox)
+        modules_path_base = '/lib/modules'
+        modules_version = self.target.kernel_version.release
+
+        if self.target.os == 'android':
+            modules_path_base = f'/vendor_dlkm{modules_path_base}'
+            try:
+                modules_version = self.target.execute(
+                    f"{busybox} ls {modules_path_base} | {busybox} head -1"
```

Comment:
`ls` is notoriously variable in its output, which makes it unsuitable for machine processing, even more so without options to restrain its behavior. `printf` and shell globbing is the most reliable way I know to list a folder with.

In the current case, is there any need for that though ? We use the `find` command anyway, so there shouldn't be any problem pointing it at one level above the one that is buggy.

Reply:
It's because if we point the already existing call to `find` one level higher it'll detect the multiple modules directories present on Linux targets. I just replaced the `ls` call with a different call to `find` which will have the same result. The options used are present in `busybox find`.

Reply:
> it'll detect the multiple modules directories present on Linux targets

I see, that's pretty hacky since we basically allow a broken setup as long as there is one folder only (but only on android)

Reply:
How do you mean? It's just because Linux targets don't usually clean their `/lib/modules` on kernel upgrades, my thinkpad has 39 different ones there. But since Linux targets keep their modules and their base kernel in the same repo the modules for the running kernel will always match `uname -r`.

On Android the two are in separate repos which if you do incremental builds can result in a mismatch but the modules partition is always swapped out so there's always only one directory there, no leftovers.

Ie on Android we don't necessarily know the name but we know there's only one directory. On Linux we know the name but there might be multiple. This implementation caters for both, nothing particularly broken about that as far as I can see.

Reply:
> Ie on Android we don't necessarily know the name but we know there's only one directory.

That's my point. There are only 2 sane behaviors:
* like what most distro do, i.e. a folder per kernel (`/lib/modules/$(uname -r)`)
* Support only a single kernel, and therefore do not bother with the `uname -r` level

Having a single level named as the `uname -r` output of another kernel is not sane in any way. AFAICU from what you said that is the situation on android.

This situation is indistinguishable from a Linux setup that would simply have no module at all installed for the current kernel, and a leftover modules folder from another kernel. In that situation, it would be completely pointless to try to load anything found there, since we know it does not match the current kernel. We only try to do that because it's android, and we know that the build system is basically broken.

Reply:
Yeah that's why we only do this 'trick' after checking if we're working with an Android target.
It's not necessarily broken, because the KMI is frozen the Android modules will work fine even against different base kernel versions for the most part, that's intentional. If someone is in the position to try to load the Lisa module that means that the phone booted which means that the setup works - it wouldn't boot without the modules being compatible. The main reason for the mismatch for us is that if you're rebuilding the kernels you probably want them tagged with the git hash which obviously changes every commit even if none of the changes affect the modules.
This is just what the environment we work with is, at least it's a problem that's pretty simple to solve.

The GKI itself comes prebuilt by default so pretty much every even stock Android setup will have a mismatch between their modules and the base kernel to begin with.

Reply:
> It's not necessarily broken, because the KMI is frozen

It is, if the modules are tied to a KMI more than a specific kernel name they could use a hash of the ABI. Any way you slice it, that is a dirty side of that build system. Yes we can work around, yes we can explain it, does not make it right. It's just like this: https://github.com/ARM-software/lisa/blob/150f9ed4e3adc3f62847126f880813491af6f412/lisa/_kmod.py#L968

I'm wondering how it works inside android itself, since that obviously breaks modprobe, but maybe they have a patched version that ignores the kernel name (or they don't use modprobe at all).

> at least it's a problem that's pretty simple to solve.

Yes somewhat. That opens the door to trying to load something knowingly broken (in case the module is actually not compatible) since we can't distinguish that anymore, but we have to be able to deal with load failure anyway.  It turns out that busybox rmmod does support -f but only if compiled with !ENABLE_MODPROBE_SMALL so I'll check that's what we have in devlib

Reply:
> That opens the door to trying to load something knowingly broken

This is still a non-issue though, this scenario pretty much cannot happen. There's only one set of modules on Android and they are necessary for the system to work - this includes the scheduler vendor hooks and even the old Lisa module they have included as `sched_tp`. If someone is in a position to load the Lisa module this means that the system has booted so the modules have all been loaded and did not crash. Pondering the hypothetical incompatibilities is all well and good but this isn't actually a real problem.

The worst case scenario is that the module fails to load and the out-of-tree build starts regardless. Even then for someone to get there all of the following would need to be true:
* the modules are similar enough to the old version for the incremental build to not trigger a rebuild
* the modules are similar enough for every single system module to load on boot and not crash the boot
* the modules are different enough for specifically the Lisa module to fail to load in some way while the `vh_sched` and `sched_tp` modules using **the same internal sched headers** work completely fine

All in all, very unlikely if not close to impossible.

Reply:
Or be working with a pre-GKI or post-this-build-system-nonsense kernel :) Here we don't have a choice and it's the most sensible way forward, but that's definitely the sort of cruft that end up creating issues (if only by creating an unmaintained path the day this problem is fixed in android).

> while the vh_sched and sched_tp modules using the same internal sched headers work completely fine

Out of curiosity, are these modules systematically loaded or are they just available for someone to modprobe dynamically ? I can't imagine they ship production system with those enabled.

Reply:
> That's very optimistic I'd say haha..

I consider "fixed" a problem where the original problematic software got replaced wholesale by something else :) 

> The sched_tp module is so tightly integrated into their scheduler hooks that Google's Android scheduler doesn't work without it.

I wish I didn't asked ...